src/main/java/net/snowflake/ingest/streaming/internal/SubscopedTokenExternalVolumeManager.java

```diff
@@ -0,0 +1,230 @@
+        volume.getFileLocationInfo().getPath(),
+        Utils.getTwoHexChars(),
+        this.counter.getAndIncrement());
+  }
+
+  @VisibleForTesting
+  static BlobPath generateBlobPathFromLocationInfoPath(
+      String fullyQualifiedTableName,
+      String filePathRelativeToVolume,
+      String twoHexChars,
+      long counterValue) {
+    String[] parts = filePathRelativeToVolume.split("/");
+    if (parts.length < 5) {
+      logger.logError(
+          "Invalid file path returned by server. Table=%s FilePathRelativeToVolume=%s",
+          fullyQualifiedTableName, filePathRelativeToVolume);
+      throw new SFException(ErrorCode.INTERNAL_ERROR, "File path returned by server is invalid");
+    }
+
+    // add twoHexChars as a prefix to the fileName (the last part of fileLocationInfo.getPath)
+    String fileNameRelativeToCredentialedPath = parts[parts.length - 1];
+    fileNameRelativeToCredentialedPath =
+        String.join("/", twoHexChars, fileNameRelativeToCredentialedPath);
+
+    // set this new fileName (with the prefix) back on the parts array so the full path can be
+    // reconstructed
+    parts[parts.length - 1] = fileNameRelativeToCredentialedPath;
+    filePathRelativeToVolume = String.join("/", parts);
+
+    // add a monotonically increasing counter at the end and the file extension
+    String suffix = counterValue + ".parquet";
+
+    return new BlobPath(
+        fileNameRelativeToCredentialedPath + suffix /* uploadPath */,
+        filePathRelativeToVolume + suffix /* fileRegistrationPath */);
+  }
+
+  /**
+   * Get the client prefix from first external volume in the map
+   *
+   * @return the client prefix
+   */
+  @Override
+  public String getClientPrefix() {
+    return this.clientPrefix;
+  }
+
+  @Override
+  public FileLocationInfo getRefreshedLocation(TableRef tableRef, Optional<String> fileName) {
+    try {
+      RefreshTableInformationResponse response =
+          this.serviceClient.refreshTableInformation(
+              new RefreshTableInformationRequest(tableRef, this.role, true));
+      logger.logDebug("Refreshed tokens for table=%s", tableRef);
+      if (response.getIcebergLocationInfo() == null) {
+        logger.logError(
+            "Did not receive location info, this will cause ingestion to grind to a halt."
+                + " TableRef=%s");
+      } else {
+        Map<String, String> creds = response.getIcebergLocationInfo().getCredentials();
+        if (creds == null || creds.isEmpty()) {
+          logger.logError(
+              "Did not receive creds in location info, this will cause ingestion to grind to a"
+                  + " halt. TableRef=%s");
+        }
+      }
+
+      return response.getIcebergLocationInfo();
+    } catch (IngestResponseException | IOException e) {
+      throw new SFException(e, ErrorCode.REFRESH_TABLE_INFORMATION_FAILURE, e.getMessage());
+    }
+  }
+
+  private InternalStage getVolumeSafe(String fullyQualifiedTableName) {
```

Comment:
What is `Safe` about it? Consider renaming.

Reply:
I believe this is just a snowflake service side convention to have a method pair for situations where the result can be null.

For instance you can see a lot of `RetVal getField()` and `RetVal getFieldSafe()` method pairs, where the former returns null when field is null, whereas the latter throws if it only has a null to return back.

I prefer an Optional<T> return value to indicate a nullable return type, but this pattern is already imprinted across snowflake so am just carrying it along!


In this case, getVolume() can technically return null thus the *Safe suffix to indicate the method will throw if it can't find an object to return back.